tensilelite/rocisa/rocisa/src/instruction/common.cpp

```diff
@@ -0,0 +1,1963 @@
+/* ************************************************************************
+ * Copyright (C) 2025 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * ************************************************************************ */
+#include "instruction/common.hpp"
+
+#include <nanobind/nanobind.h>
+#include <nanobind/stl/optional.h>
+#include <nanobind/stl/shared_ptr.h>
+#include <nanobind/stl/string.h>
+#include <nanobind/stl/variant.h>
+
+namespace nb = nanobind;
+
+void common_inst(nb::module_ m_common)
+{
+    nb::class_<rocisa::SAbsI32, rocisa::CommonInstruction>(m_common, "SAbsI32")
+        .def(nb::init<const std::shared_ptr<rocisa::RegisterContainer>&,
+                      const std::shared_ptr<rocisa::RegisterContainer>&,
+                      const std::string&>(),
+             nb::arg("dst"),
+             nb::arg("src"),
+             nb::arg("comment") = "")
+        .def("__deepcopy__", [](const rocisa::SAbsI32& self, const nb::dict&) {
+            return new rocisa::SAbsI32(self);
+        });
+
+    nb::class_<rocisa::SMaxI32, rocisa::CommonInstruction>(m_common, "SMaxI32")
```

Comment:
The declarations here look boiler plate, could we use a macro to reduce the number of lines? Something like
```c++
#define ROCISA_DECLARE_DST_SRC(INSTRUCTION) \
  nb::class_<rocisa::INSTRUCTION, rocisa::CommonInstruction>(m_common, TOSTRING(INSTRUCTION)) \
  ... \
  .. return new rocisa::INSTRUCTION(self); });  

#define ROCISA_DECLARE_DST_SRC0_SRC1(INSTRUCTION) ..

ROCISA_DECLARE_DST_SRC(SAbsI32);
ROCISA_DECLARE_DST_SRC0_SRC1(SMaxI32);
ROCISA_DECLARE_DST_SRC0_SRC1(SMaxU32);
...
```
Similar for the declaration in headers, and for other types (cvt, instruction, mem, etc..)

Reply:
It'll be a little complicated since there might be an exception. Or by that time we'll choose not to use macro instead? (Actually I'm not a fan of a macro because if compiler error occurs, the line happens in the macro)

Reply:
For instructions where there is an exception to the declaration, we can just use not use the macro. I'm guessing a lot of these declarations will be similar though.

> if compiler error occurs, the line happens in the macro

Fair point. There are definitely trade-offs with using macros.